addons/base_vat/models/res_partner.py

```diff
@@ -168,27 +176,30 @@ def check_vat(self):
                 msg = partner._build_vat_error_message(country and country.code.lower() or None, partner.vat, partner_label)
                 raise ValidationError(msg)
 
-    @api.onchange('vat', 'country_id')
-    def _onchange_check_vies(self):
-        """ Check the VAT number with VIES, if enabled. Return a non-blocking warning if the check fails."""
-        if self.env.context.get('company_id'):
-            company = self.env['res.company'].browse(self.env.context['company_id'])
-        else:
-            company = self.env.company
-        if not company.vat_check_vies:
-            return
-
-        eu_countries = self.env.ref('base.europe').country_ids
-        for eu_partner_company in self.filtered(lambda partner: partner.country_id in eu_countries and partner.is_company):
+    @api.depends('vat', 'country_id')
+    def _compute_vies_valid(self):
+        """ Check the VAT number with VIES, if enabled."""
+        for partner in self:
+            if not partner.perform_vies_check:
+                continue
```

Comment:
This doesn't work because `perform_vies_check` depends on the context.
For instance:
* Company BE has `vat_check_vies=True`
* Company US has `vat_check_vies=False`
* Company BE creates a new (shared) partner with a valid VAT
* It is checked here, the fields is set `vies_valid=True`
* Company US changes it for whatever reason to something not valid
* The field is not updated
* Company BE still sees this as `vies_valid=True`

You'd have the same kind of issue if you set  `vies_valid=False` when `perform_vies_check==False`


So I'd say: if the partner is linked to a company, then you can compute `perform_vies_check` based on that company. If it isn't, then you need to check all the companies, and if one of them needs to check VIES, do it for all of them.

Reply:
yes, that makes it very clear :+1: